{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6de4c586",
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import defaultdict\n",
    "from Bio.SeqIO.FastaIO import SimpleFastaParser as sfp\n",
    "from Bio import SeqIO, SearchIO, Entrez\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import os, glob, math, re, gzip, skbio, time\n",
    "import matplotlib.pyplot as plt\n",
    "import warnings\n",
    "from Bio import Entrez\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d1ea93e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def cmdir(path):\n",
    "    if not os.path.isdir(path):\n",
    "        os.mkdir(path)\n",
    "\n",
    "def scaffold(gene):\n",
    "    if gene != \"None\":\n",
    "        try: return re.search(\"(.+?)_[0-9]+$\", gene).group(1)\n",
    "        except: print(gene)\n",
    "\n",
    "def sbatch(name, cpus, cmd):\n",
    "    return \"sbatch -J %s -p serc -t 1- -c %d --mem %dG --wrap '%s'\" %(name, cpus, cpus*8, cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ddaa3368",
   "metadata": {},
   "outputs": [],
   "source": [
    "rootdir = \"/scratch/users/ajaffe/photoeco/\"\n",
    "cmdir(rootdir + \"curate\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46fb9927",
   "metadata": {},
   "source": [
    "### read in data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cb5086d3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_name</th>\n",
       "      <th>lineage</th>\n",
       "      <th>checkm_completeness</th>\n",
       "      <th>checkm_redundancy</th>\n",
       "      <th>cluster95</th>\n",
       "      <th>cluster99</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2626541517</td>\n",
       "      <td>AMZ IB</td>\n",
       "      <td>56.79</td>\n",
       "      <td>6.11</td>\n",
       "      <td>5_1</td>\n",
       "      <td>5_5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2626541518</td>\n",
       "      <td>AMZ IB</td>\n",
       "      <td>60.34</td>\n",
       "      <td>6.90</td>\n",
       "      <td>5_1</td>\n",
       "      <td>5_11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2626541532</td>\n",
       "      <td>AMZ IB</td>\n",
       "      <td>65.62</td>\n",
       "      <td>2.72</td>\n",
       "      <td>5_1</td>\n",
       "      <td>5_15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2626541536</td>\n",
       "      <td>AMZ IB</td>\n",
       "      <td>50.14</td>\n",
       "      <td>5.10</td>\n",
       "      <td>5_1</td>\n",
       "      <td>5_21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2818991511</td>\n",
       "      <td>AMZ II</td>\n",
       "      <td>74.64</td>\n",
       "      <td>0.27</td>\n",
       "      <td>4_1</td>\n",
       "      <td>4_4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  genome_name lineage  checkm_completeness  checkm_redundancy cluster95  \\\n",
       "0  2626541517  AMZ IB                56.79               6.11       5_1   \n",
       "1  2626541518  AMZ IB                60.34               6.90       5_1   \n",
       "2  2626541532  AMZ IB                65.62               2.72       5_1   \n",
       "3  2626541536  AMZ IB                50.14               5.10       5_1   \n",
       "4  2818991511  AMZ II                74.64               0.27       4_1   \n",
       "\n",
       "  cluster99  \n",
       "0       5_5  \n",
       "1      5_11  \n",
       "2      5_15  \n",
       "3      5_21  \n",
       "4       4_4  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genomes = pd.read_csv(rootdir + \"genome_table.tsv\", sep=\"\\t\")\n",
    "genomes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71dd2c53",
   "metadata": {},
   "source": [
    "### run clustering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "19c620f9",
   "metadata": {},
   "outputs": [],
   "source": [
    "prodigal = open(rootdir + \"reassembly/genomes/prodigal.sh\", \"w\")\n",
    "\n",
    "for genome in glob.glob(rootdir + \"reassembly/genomes/*fna\"):\n",
    "   \n",
    "    name = os.path.basename(genome).replace(\".fna\",\"\")\n",
    "    #gpath = rootdir + \"recover/bins/\" + name + \".fa\"\n",
    "    \n",
    "    #newname = rootdir + \"recover/genomes/tmp/\" + name + \".fna\"\n",
    "    #with open(newname, \"w\") as out:\n",
    "    #    for record in sfp(open(gpath)):\n",
    "    #        out.write(\">%s\\n%s\\n\" %(record[0], record[1]))\n",
    "    prodigal.write(\"prodigal -i %s -a %s -p single\\n\" %(genome, genome.replace(\".fna\", \".genes.faa\")))\n",
    "\n",
    "prodigal.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "30eb006b",
   "metadata": {},
   "outputs": [],
   "source": [
    "wrapper = open(rootdir + \"curate/cluster.sh\", \"w\")\n",
    "orf2bin = {}\n",
    "\n",
    "for lineage in [\"AMZ IA\", \"AMZ IB\", \"AMZ II\", \"AMZ III\"]:\n",
    "    \n",
    "    workdir = rootdir + \"curate/\" + lineage.lower().replace(\" \", \"\")\n",
    "    cmdir(workdir)\n",
    "    \n",
    "    table = genomes[genomes[\"lineage\"]==lineage]\n",
    "    \n",
    "    # write the proteins\n",
    "    with open(workdir + \"/lineage.faa\", \"w\") as out:\n",
    "        for key, row in table.iterrows():\n",
    "            if \"RR\" in row[\"genome_name\"]:\n",
    "                faapath = rootdir + \"reassembly/genomes/%s.genes.faa\" %(row[\"genome_name\"])\n",
    "            else:\n",
    "                faapath = rootdir + \"genomes/filtered/%s.genes.faa\" %(row[\"genome_name\"])\n",
    "            for record in sfp(open(faapath)):\n",
    "                out.write(\">%s\\n%s\\n\" %(record[0].split(\" # \")[0], record[1]))\n",
    "                orf2bin[record[0].split(\" # \")[0]] = row[\"genome_name\"]\n",
    "    \n",
    "    # cluster the proteins\n",
    "    makedb = \"mmseqs createdb %s %s\" %(workdir + \"/lineage.faa\", workdir + \"/lineage.db\")\n",
    "    cluster = \"mmseqs cluster --cov-mode 0 --threads 4 %s %s %s\" %(workdir + \"/lineage.db\", \n",
    "        workdir + \"/lineage.cluster\", workdir + \"/tmp\")\n",
    "    process = \"mmseqs createtsv %s %s %s %s\" %(workdir + \"/lineage.db\", workdir + \"/lineage.db\",\n",
    "                                              workdir + \"/lineage.cluster\", workdir + \"/lineage.cluster.tsv\")\n",
    "    wrapper.write(\"\\n\".join([makedb, cluster, process]))\n",
    "    wrapper.write(\"\\n\")\n",
    "    \n",
    "wrapper.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64818ee1",
   "metadata": {},
   "source": [
    " ### analyze"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "778b1d81",
   "metadata": {},
   "outputs": [],
   "source": [
    "results = []\n",
    "\n",
    "for result in glob.glob(rootdir + \"curate/amz*/lineage.cluster.tsv\"):\n",
    "    \n",
    "    table = pd.read_csv(result, sep=\"\\t\", header=None)\n",
    "    table.columns = [\"centroid\", \"gene\"]\n",
    "    table[\"scaffold\"] = table[\"gene\"].apply(lambda x: \"_\".join(x.split(\"_\")[:-1]))\n",
    "    table[\"bin\"] = table[\"gene\"].map(orf2bin)\n",
    "    table[\"lineage\"] = result.split(\"/\")[-2]\n",
    "    \n",
    "    centroids = {}\n",
    "    \n",
    "    for i, centroid in enumerate(table.centroid.unique()):\n",
    "        centroids[centroid] = i\n",
    "    \n",
    "    table[\"cluster\"] = table[\"centroid\"].map(centroids)\n",
    "    results.append(table)\n",
    "\n",
    "all_results = pd.concat(results)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "33c158ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>centroid</th>\n",
       "      <th>gene</th>\n",
       "      <th>scaffold</th>\n",
       "      <th>bin</th>\n",
       "      <th>lineage</th>\n",
       "      <th>cluster</th>\n",
       "      <th>perc_total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR599070.100.50_contig_114_3</td>\n",
       "      <td>ERR599070.100.50_contig_114_3</td>\n",
       "      <td>ERR599070.100.50_contig_114</td>\n",
       "      <td>ERR599070.100.50</td>\n",
       "      <td>amzia</td>\n",
       "      <td>0</td>\n",
       "      <td>12.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR599070.100.50_contig_114_3</td>\n",
       "      <td>JASLWC010000035.1_14</td>\n",
       "      <td>JASLWC010000035.1</td>\n",
       "      <td>GCA_030741055.1_ASM3074105v1_genomic</td>\n",
       "      <td>amzia</td>\n",
       "      <td>0</td>\n",
       "      <td>12.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        centroid                           gene  \\\n",
       "0  ERR599070.100.50_contig_114_3  ERR599070.100.50_contig_114_3   \n",
       "1  ERR599070.100.50_contig_114_3           JASLWC010000035.1_14   \n",
       "\n",
       "                      scaffold                                   bin lineage  \\\n",
       "0  ERR599070.100.50_contig_114                      ERR599070.100.50   amzia   \n",
       "1            JASLWC010000035.1  GCA_030741055.1_ASM3074105v1_genomic   amzia   \n",
       "\n",
       "   cluster  perc_total  \n",
       "0        0        12.5  \n",
       "1        0        12.5  "
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pfams = all_results.groupby([\"lineage\", \"cluster\"], as_index=False).aggregate({\"bin\":\"nunique\"})\n",
    "genomes[\"lineage_brief\"] = genomes[\"lineage\"].apply(lambda x: x.lower().replace(\" \", \"\"))\n",
    "totals = genomes.groupby(\"lineage_brief\", as_index=False).aggregate({\"genome_name\":\"nunique\"}).rename(columns={\"genome_name\":\"total\"})\n",
    "pfams = pfams.merge(totals, how=\"left\", left_on=\"lineage\", right_on=\"lineage_brief\").drop(\"lineage_brief\", axis=1)\n",
    "pfams[\"perc_total\"] = pfams.apply(lambda x: (x[\"bin\"]/x[\"total\"])*100, axis=1)\n",
    "all_results = all_results.merge(pfams[[\"lineage\", \"cluster\", \"perc_total\"]], how=\"left\")\n",
    "all_results.head(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7e4486c",
   "metadata": {},
   "source": [
    "### trim"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "d37a414b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>lineage</th>\n",
       "      <th>bin</th>\n",
       "      <th>scaffold</th>\n",
       "      <th>num_pfams</th>\n",
       "      <th>num_rare</th>\n",
       "      <th>perc_rare</th>\n",
       "      <th>trim</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>amzia</td>\n",
       "      <td>ERR599070.100.50</td>\n",
       "      <td>ERR599070.100.50_contig_114</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>amzia</td>\n",
       "      <td>GCA_030741055.1_ASM3074105v1_genomic</td>\n",
       "      <td>JASLWC010000035.1</td>\n",
       "      <td>14</td>\n",
       "      <td>1</td>\n",
       "      <td>7.142857</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>amzia</td>\n",
       "      <td>ERR598987.100.72</td>\n",
       "      <td>ERR598987.100.72_scaffold_149</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>33.333333</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>amzia</td>\n",
       "      <td>ERR599070.100.50</td>\n",
       "      <td>ERR599070.100.50_contig_192</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>amzia</td>\n",
       "      <td>ERR599070.100.50</td>\n",
       "      <td>ERR599070.100.50_contig_196</td>\n",
       "      <td>23</td>\n",
       "      <td>2</td>\n",
       "      <td>8.695652</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  lineage                                   bin  \\\n",
       "0   amzia                      ERR599070.100.50   \n",
       "1   amzia  GCA_030741055.1_ASM3074105v1_genomic   \n",
       "2   amzia                      ERR598987.100.72   \n",
       "3   amzia                      ERR599070.100.50   \n",
       "4   amzia                      ERR599070.100.50   \n",
       "\n",
       "                        scaffold  num_pfams  num_rare  perc_rare   trim  \n",
       "0    ERR599070.100.50_contig_114          4         2  50.000000   True  \n",
       "1              JASLWC010000035.1         14         1   7.142857  False  \n",
       "2  ERR598987.100.72_scaffold_149          6         2  33.333333  False  \n",
       "3    ERR599070.100.50_contig_192          5         1  20.000000  False  \n",
       "4    ERR599070.100.50_contig_196         23         2   8.695652  False  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "trim_info = defaultdict(list)\n",
    "\n",
    "for lineage in all_results.lineage.unique():\n",
    "    \n",
    "    table = all_results[all_results[\"lineage\"]==lineage]\n",
    "    \n",
    "    for scaffold in table.scaffold.unique():\n",
    "        \n",
    "        subtable = table[table[\"scaffold\"]==scaffold].drop_duplicates(\"cluster\")\n",
    "        rare = len(subtable.query(\"perc_total<10\"))\n",
    "        rare_perc = rare/len(subtable) * 100\n",
    "        \n",
    "        trim_info[\"lineage\"].append(lineage)\n",
    "        trim_info[\"bin\"].append(subtable[\"bin\"].iloc[0])\n",
    "        trim_info[\"scaffold\"].append(scaffold)\n",
    "        trim_info[\"num_pfams\"].append(len(subtable))\n",
    "        trim_info[\"num_rare\"].append(rare)\n",
    "        trim_info[\"perc_rare\"].append(rare_perc)\n",
    "        trim_info[\"trim\"].append(rare_perc>=50)\n",
    "\n",
    "trimdf = pd.DataFrame(trim_info)\n",
    "trimdf.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b5094b7",
   "metadata": {},
   "source": [
    "### export"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "9376e81b",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmdir(rootdir + \"curate/trimmed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "ff00b8fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "for bin in trimdf.bin.unique():\n",
    "    \n",
    "    # only do it for newly resolved\n",
    "    if \"RR\" in bin:\n",
    "        \n",
    "        totrim = trimdf[(trimdf[\"bin\"]==bin) & \\\n",
    "                        (trimdf[\"trim\"]==True)][\"scaffold\"].to_list()\n",
    "\n",
    "        gpath = rootdir + \"reassembly/genomes/%s.fna\" %(bin)\n",
    "\n",
    "        with open(rootdir + \"curate/trimmed/\" + bin + \".fna\", \"w\") as out:\n",
    "            for record in sfp(open(gpath)):\n",
    "                scaf = record[0].split(\" \")[0]\n",
    "                if scaf not in totrim:\n",
    "                    out.write(\">%s\\n%s\\n\" %(scaf, record[1]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "93f2f7fd",
   "metadata": {},
   "source": [
    "### compare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "7f428b84",
   "metadata": {},
   "outputs": [],
   "source": [
    "cmdir(rootdir + \"curate/quality\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "817eec12",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "checkm qa -t 20 -o 1 -f /scratch/users/ajaffe/photoeco/curate/quality/output_table.txt --tab_table /scratch/users/ajaffe/photoeco/curate/quality/lineage.ms /scratch/users/ajaffe/photoeco/curate/quality/\n"
     ]
    }
   ],
   "source": [
    " # construct checkm calls\n",
    "call = \"checkm lineage_wf -t 20 -x .fna --pplacer_threads 20 %s %s\" %(rootdir + \"curate/trimmed\", rootdir + \"curate/quality\")\n",
    "call2 = \"checkm qa -t 20 -o 1 -f %s --tab_table %s %s\" %(rootdir + \"curate/quality/output_table.txt\",\n",
    "    rootdir + \"curate/quality/lineage.ms\", rootdir + \"curate/quality/\")\n",
    "print(call2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "2f179bd0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_name</th>\n",
       "      <th>new_completeness</th>\n",
       "      <th>new_redundancy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR598946.10.10</td>\n",
       "      <td>59.98</td>\n",
       "      <td>0.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ERR598946.100.35</td>\n",
       "      <td>83.70</td>\n",
       "      <td>1.45</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        genome_name  new_completeness  new_redundancy\n",
       "0   ERR598946.10.10             59.98            0.54\n",
       "1  ERR598946.100.35             83.70            1.45"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in new results\n",
    "checkm_df = pd.read_csv(rootdir + \"curate/quality/output_table.txt\", sep=\"\\t\").reset_index()\n",
    "checkm_df = checkm_df[[\"Bin Id\", \"Completeness\", \"Contamination\"]]\n",
    "checkm_df.columns = [\"genome_name\", \"new_completeness\", \"new_redundancy\"]\n",
    "checkm_df.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a6baa859",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome_name</th>\n",
       "      <th>new_completeness</th>\n",
       "      <th>new_redundancy</th>\n",
       "      <th>checkm_completeness</th>\n",
       "      <th>checkm_redundancy</th>\n",
       "      <th>delta_completeness</th>\n",
       "      <th>delta_redundancy</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>SRR11923210.100.2</td>\n",
       "      <td>37.90</td>\n",
       "      <td>0.32</td>\n",
       "      <td>56.65</td>\n",
       "      <td>1.13</td>\n",
       "      <td>-18.75</td>\n",
       "      <td>-0.81</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>SRR13782137.100.27</td>\n",
       "      <td>39.65</td>\n",
       "      <td>1.38</td>\n",
       "      <td>61.07</td>\n",
       "      <td>5.09</td>\n",
       "      <td>-21.42</td>\n",
       "      <td>-3.71</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>SRR16028415.100.26</td>\n",
       "      <td>45.97</td>\n",
       "      <td>2.70</td>\n",
       "      <td>61.92</td>\n",
       "      <td>7.05</td>\n",
       "      <td>-15.95</td>\n",
       "      <td>-4.35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ERR868457.100.2</td>\n",
       "      <td>50.56</td>\n",
       "      <td>2.85</td>\n",
       "      <td>63.06</td>\n",
       "      <td>4.44</td>\n",
       "      <td>-12.50</td>\n",
       "      <td>-1.59</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>SRR25584961.100.55</td>\n",
       "      <td>52.43</td>\n",
       "      <td>0.74</td>\n",
       "      <td>69.37</td>\n",
       "      <td>1.69</td>\n",
       "      <td>-16.94</td>\n",
       "      <td>-0.95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>SRR13782129.100.144</td>\n",
       "      <td>55.54</td>\n",
       "      <td>0.95</td>\n",
       "      <td>55.54</td>\n",
       "      <td>0.95</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>SRR11923215.100.61</td>\n",
       "      <td>59.65</td>\n",
       "      <td>0.82</td>\n",
       "      <td>67.62</td>\n",
       "      <td>1.36</td>\n",
       "      <td>-7.97</td>\n",
       "      <td>-0.54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ERR598946.10.10</td>\n",
       "      <td>59.98</td>\n",
       "      <td>0.54</td>\n",
       "      <td>76.74</td>\n",
       "      <td>3.14</td>\n",
       "      <td>-16.76</td>\n",
       "      <td>-2.60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>SRR4465025.100.54</td>\n",
       "      <td>61.73</td>\n",
       "      <td>3.08</td>\n",
       "      <td>62.27</td>\n",
       "      <td>3.35</td>\n",
       "      <td>-0.54</td>\n",
       "      <td>-0.27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SRR11787837.100.56</td>\n",
       "      <td>65.11</td>\n",
       "      <td>1.24</td>\n",
       "      <td>80.65</td>\n",
       "      <td>3.68</td>\n",
       "      <td>-15.54</td>\n",
       "      <td>-2.44</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            genome_name  new_completeness  new_redundancy  \\\n",
       "23    SRR11923210.100.2             37.90            0.32   \n",
       "33   SRR13782137.100.27             39.65            1.38   \n",
       "34   SRR16028415.100.26             45.97            2.70   \n",
       "15      ERR868457.100.2             50.56            2.85   \n",
       "50   SRR25584961.100.55             52.43            0.74   \n",
       "32  SRR13782129.100.144             55.54            0.95   \n",
       "29   SRR11923215.100.61             59.65            0.82   \n",
       "0       ERR598946.10.10             59.98            0.54   \n",
       "53    SRR4465025.100.54             61.73            3.08   \n",
       "18   SRR11787837.100.56             65.11            1.24   \n",
       "\n",
       "    checkm_completeness  checkm_redundancy  delta_completeness  \\\n",
       "23                56.65               1.13              -18.75   \n",
       "33                61.07               5.09              -21.42   \n",
       "34                61.92               7.05              -15.95   \n",
       "15                63.06               4.44              -12.50   \n",
       "50                69.37               1.69              -16.94   \n",
       "32                55.54               0.95                0.00   \n",
       "29                67.62               1.36               -7.97   \n",
       "0                 76.74               3.14              -16.76   \n",
       "53                62.27               3.35               -0.54   \n",
       "18                80.65               3.68              -15.54   \n",
       "\n",
       "    delta_redundancy  \n",
       "23             -0.81  \n",
       "33             -3.71  \n",
       "34             -4.35  \n",
       "15             -1.59  \n",
       "50             -0.95  \n",
       "32              0.00  \n",
       "29             -0.54  \n",
       "0              -2.60  \n",
       "53             -0.27  \n",
       "18             -2.44  "
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in old results\n",
    "checkm_df = checkm_df.merge(genomes[[\"genome_name\", \"checkm_completeness\", \"checkm_redundancy\"]], how=\"left\")\n",
    "checkm_df[\"delta_completeness\"] = checkm_df.apply(lambda x: x[\"new_completeness\"] - x[\"checkm_completeness\"], axis=1)\n",
    "checkm_df[\"delta_redundancy\"] = checkm_df.apply(lambda x: x[\"new_redundancy\"] - x[\"checkm_redundancy\"], axis=1)\n",
    "checkm_df.sort_values(\"new_completeness\", ascending=True)[0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "fe4961da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-3.434912280701755"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "checkm_df[\"delta_completeness\"].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "03ded0e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-0.7759649122807019"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "checkm_df[\"delta_redundancy\"].mean()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
